src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -609,6 +613,24 @@ private void verifyTypeConstraints(String type, String namespace, String name) t
         }
     }
 
+    private Map<String, String> parseQualifiers(final Map<String, String> qualifiers) throws MalformedPackageURLException {
+        if (qualifiers == null || qualifiers.isEmpty()) {
+            return null;
+        }
+
+        try {
+            final TreeMap<String, String> results = qualifiers.entrySet().stream()
+                    .filter(entry -> entry.getValue() != null && !entry.getValue().isEmpty())
```

Comment:
This will silently drop invalid keys. Shouldn't an exception be thrown instead?

Maybe it is worth changing `validateQualifiers` to accept upper-case keys and call `validateQualifiers(qualifiers)` on line 622?

Reply:
It is supposed to drop invalid keys (see below where I quote the spec).